opencog/atoms/execution/Instantiator.cc

```diff
@@ -275,7 +275,10 @@ Handle Instantiator::walk_tree(const Handle& expr)
 	HandleSeq oset_results;
 	bool changed = walk_tree(oset_results, lexpr->getOutgoingSet());
 	if (changed)
-		return Handle(createLink(t, oset_results, expr->getTruthValue()));
+	{
+		LinkPtr subl = createLink(t, oset_results, expr->getTruthValue());
+		return Handle(_as->add_atom(subl));
```

Comment:
Ah OK, yes, that makes sense.  So, several remarks:
-- You didn't really need an ImplicationLink C++ class; instead, the factory needed to be altered to put a `if classserver()isType(t, SCOPE_LINK) {createScopeLink(t, args); }` to find the variables.  Or if you want to keep it, then `classsever().isType(IMPLICATION_LINK...)` to pick up the extensional etc. links
-- The factory needs to be called ... 
-- The code needs to be changed to do lazy evaluation (and do a lazy factory as well)
-- A unit test case is needed for the new ImplicationLink so that when the lazy-eval stuff is added, the implicaitonlink is not broken.